unixfs/src/dir/builder/iter.rs

```diff
@@ -0,0 +1,416 @@
+        // BufferedTreeWriter.
+        //
+        // could also be that the HAMT shard building should start earlier, since the same
+        // heuristic can be detected *at* bufferedtreewriter. there the split would be easier, and
+        // this would "just" be a single node rendering, and not need any additional states..
+
+        /// Newtype which uses the BTreeMap<String, Leaf> as Vec<PBLink>.
+        struct BTreeMappedDir<'a> {
+            links: &'a BTreeMap<String, Leaf>,
+            data: UnixFs<'a>,
+        }
+
+        /// Newtype which represents an entry from BTreeMap<String, Leaf> as PBLink as far as the
+        /// protobuf representation goes.
+        struct EntryAsPBLink<'a>(&'a String, &'a Leaf);
+
+        impl<'a> MessageWrite for EntryAsPBLink<'a> {
+            fn get_size(&self) -> usize {
+                use quick_protobuf::sizeofs::*;
+
+                // ones are the tags
+                1 + sizeof_len(self.0.len())
+                    + 1
+                    //+ sizeof_len(WriteableCid(&self.1.link).get_size())
+                    + sizeof_len(self.1.link.to_bytes().len())
+                    + 1
+                    + sizeof_varint(self.1.total_size)
+            }
+
+            fn write_message<W: WriterBackend>(
+                &self,
+                w: &mut Writer<W>,
+            ) -> quick_protobuf::Result<()> {
+                // w.write_with_tag(10, |w| w.write_message(&WriteableCid(&self.1.link)))?;
+                w.write_with_tag(10, |w| w.write_bytes(&self.1.link.to_bytes()))?;
+                w.write_with_tag(18, |w| w.write_string(self.0.as_str()))?;
+                w.write_with_tag(24, |w| w.write_uint64(self.1.total_size))?;
+                Ok(())
+            }
+        }
+
+        struct WriteableCid<'a>(&'a Cid);
+
+        // Cid by default does not have a way to count it's length or just write it out without
+        // allocating a vector.
+        impl<'a> MessageWrite for WriteableCid<'a> {
+            fn get_size(&self) -> usize {
+                use cid::Version::*;
+                use quick_protobuf::sizeofs::*;
+
+                match self.0.version() {
+                    V0 => self.0.hash().as_bytes().len(),
+                    V1 => {
+                        let version_len = 1;
+                        let codec_len = sizeof_varint(u64::from(self.0.codec()));
+                        let hash_len = self.0.hash().as_bytes().len();
+                        version_len + codec_len + hash_len
+                    }
+                }
+            }
+
+            fn write_message<W: WriterBackend>(
+                &self,
+                w: &mut Writer<W>,
+            ) -> quick_protobuf::Result<()> {
+                use cid::Version::*;
+
+                match self.0.version() {
+                    V0 => {
+                        for b in self.0.hash().as_bytes() {
+                            w.write_u8(*b)?;
```

Comment:
can't we write the whole byte slice at once? just making sure

Reply:
sadly that is not exposed; `write_bytes` would prefix it with length varint (as often types in pb are `tag`, `len`, `bytes`).